### ISSUE TITLE
Add support for field definition with a predicate

### ISSUE THREAD

**Issue by benas** [benas/random-beans#342]
As of v3.8, in order to define a field, one should use the `FieldDefinition` API. This API provides 5 methods to define field properties (`named`, `ofType`, `inClass`, `isAnnotatedWith` and `hasModifiers`).

As a library user, this limited set of properties makes it impossible to use a custom condition on a field, for example `field.isSynthetic()`.

As a library developer, I will need to update the API each time a user asks to add a custom field property.

A better approach is to use a `Predicate<Field>` as a replacement for `FieldDefinition`. Thanks to `java.util.function.Predicate#and` and `java.util.function.Predicate#or`, multiple predicates can be combined to define a field just like `FieldDefinition`, adding the flexibility to define custom conditions **without having to change the API each time**. Here is an example:

```java
Predicate<Field> predicate = named("name").and(ofType(String.class)).and(inClass(Person.class));
```

Random Beans would provide common predicates and let the user define additional predicates as needed. A first work on this feature can be reviewed in branch [`predicates`](https://github.com/benas/random-beans/commit/aa5f0da68c439d02d264cbea0dd788de9124852b).

@PascalSchumacher WDYT?

BTW, this idea came from a situation where I wanted to exclude all types defined in a given package. I though about adding a [TypeDefinition](https://github.com/benas/random-beans/commit/4a91de9eca95cc70b238b64e91d4b70b091f58e2) API allowing one to write something like:

```java
EnhancedRandom enhancedRandom = new EnhancedRandomBuilder()
	        .exclude(TypeDefinition.type().inPackage("io.github.benas.randombeans.beans").get())
	        .build();
```

But then I found myself exposing [all properties](https://github.com/benas/random-beans/commit/9dbe236cdb474c8fb2d4edc0b9d9c3a119e53fcf) of a `Class`. I quickly realized that this was not the best approach.. Hence the idea of using predicates: No need to expose properties + ability to provide custom conditions.

**Comment by PascalSchumacher** [comment]
Good idea. 👍

**Comment by benas** [comment]
@PascalSchumacher Thank you for your quick feedback!

Rebased and merged as 4929f12a5a3c832fa8770935d5542deb2a9482ac.
